// @adonisjs/hash/build/src/Drivers/Fake.js
"use strict";
/*
 * @adonisjs/hash
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fake = void 0;
/**
 * Generates and verifies hash using no algorigthm.
 */
class Fake {
    constructor() {
        this.ids = ['fake'];
    }
    /**
     * Alias for [[this.make]]
     */
    hash(value) {
        return Promise.resolve(value);
    }
    /**
     * Returns hash for a given value
     */
    make(value) {
        return this.hash(value);
    }
    /**
     * Verify hash to know if two values are same.
     */
    verify(hashedValue, plainValue) {
        return Promise.resolve(hashedValue === plainValue);
    }
    /**
     * Returns a boolean telling if hash needs a rehash. Returns true when
     * one of the original params have been changed.
     */
    needsReHash(_value) {
        return false;
    }
}
exports.Fake = Fake;
