// create-adonis-ts-app/build/Tasks/Scaffold/createEditorConfig.js
"use strict";
/*
 * adonis-ts-boilerplate
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const sink_1 = require("@adonisjs/sink");
/**
 * Create `.editorconfig` inside destination
 */
const task = (_, logger, { absPath }) => {
    const editorConfig = new sink_1.files.IniFile(absPath, '.editorconfig');
    /**
     * All files
     */
    editorConfig.set('*', {
        indent_style: 'space',
        indent_size: 2,
        end_of_line: 'lf',
        charset: 'utf-8',
        trim_trailing_whitespace: true,
        insert_final_newline: true,
    });
    /**
     * JSON file
     */
    editorConfig.set('*.json', {
        insert_final_newline: 'ignore',
    });
    /**
     * Markdown files
     */
    editorConfig.set('*.md', {
        trim_trailing_whitespace: false,
    });
    editorConfig.commit();
    logger.action('create').succeeded('.editorconfig');
};
exports.default = task;
